File: c/two_bucket.c
#include "two_bucket.h"

#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct list_node {
  bucket_liters_t b1;
  bucket_liters_t b2;
  int count;
  struct list_node *next;
} list_node_t;

typedef struct {
  list_node_t *head;
  list_node_t *tail;
} queue_t;

queue_t *create_queue();
bool queue_is_empty(queue_t *q);
void dequeue(queue_t *q, bucket_liters_t *b1, bucket_liters_t *b2, int *count);
void enqueue(queue_t *q, bucket_liters_t b1, bucket_liters_t b2, int count);
void free_queue(queue_t *q);
bool queue_contains(queue_t *q, bucket_liters_t b1, bucket_liters_t b2);
char *strdup(const char *s);

bucket_result_t measure(bucket_liters_t bucket_1_size,
                        bucket_liters_t bucket_2_size,
                        bucket_liters_t goal_volume, bucket_id_t start_bucket) {
  bucket_result_t res;
  queue_t *q = create_queue();
  queue_t *set = create_queue();
  bucket_liters_t b1 = 0, b2 = 0;
  int count = 0;
  enqueue(q, b1, b2, count);
  while (!queue_is_empty(q)) {
    dequeue(q, &b1, &b2, &count);
    if (queue_contains(set, b1, b2)) {
      continue;
    }
    enqueue(set, b1, b2, 0);
    if (start_bucket == BUCKET_ID_1) {
      if (b1 == 0 && b2 == bucket_2_size) {
        continue;
      }
    } else {
      if (b1 == bucket_1_size && b2 == 0) {
        continue;
      }
    }
    if (b1 == goal_volume) {
      res.possible = true;
      res.goal_bucket = BUCKET_ID_1;
      res.move_count = count;
      res.other_bucket_liters = b2;
      goto RET;
    }
    if (b2 == goal_volume) {
      res.possible = true;
      res.goal_bucket = BUCKET_ID_2;
      res.move_count = count;
      res.other_bucket_liters = b1;
      goto RET;
    }
    bucket_liters_t diff = fmin(bucket_2_size - b2, b1);
    enqueue(q, b1 - diff, b2 + diff, count + 1);
    enqueue(q, 0, b2, count + 1);
    enqueue(q, bucket_1_size, b2, count + 1);
    diff = fmin(bucket_1_size - b1, b2);
    enqueue(q, b1 + diff, b2 - diff, count + 1);
    enqueue(q, b1, 0, count + 1);
    enqueue(q, b1, bucket_2_size, count + 1);
  }

  res.possible = false;

RET:
  free_queue(q);
  free_queue(set);
  return res;
}

queue_t *create_queue() {
  queue_t *q = malloc(sizeof(queue_t));
  if (q == NULL) {
    return NULL;
  }

  q->head = NULL;
  q->tail = NULL;
  return q;
}

bool queue_is_empty(queue_t *q) { return q->head == NULL; }

void dequeue(queue_t *q, bucket_liters_t *b1, bucket_liters_t *b2, int *count) {
  list_node_t *node = q->head;
  q->head = node->next;
  if (q->head == NULL) {
    q->tail = NULL;
  }

  *b1 = node->b1;
  *b2 = node->b2;
  *count = node->count;
  free(node);
}

void enqueue(queue_t *q, bucket_liters_t b1, bucket_liters_t b2, int count) {
  list_node_t *node = malloc(sizeof(list_node_t));
  if (node == NULL) {
    return;
  }
  node->b1 = b1;
  node->b2 = b2;
  node->count = count;
  node->next = NULL;

  if (q->tail == NULL) {
    q->head = node;
    q->tail = node;
  } else {
    q->tail->next = node;
    q->tail = node;
  }
}

void free_queue(queue_t *q) {
  while (q->head != NULL) {
    list_node_t *node = q->head;
    q->head = q->head->next;
    free(node);
  }
  free(q);
  q = NULL;
}

bool queue_contains(queue_t *q, bucket_liters_t b1, bucket_liters_t b2) {
  list_node_t *curr = q->head;
  while (curr != NULL) {
    if (curr->b1 == b1 && curr->b2 == b2) {
      return true;
    }
    curr = curr->next;
  }
  return false;
}

File: lua/two-bucket.lua
return {
    measure = function(args)
        local q = {}
        local st = {}
        table.insert(q, { 0, 0, 0 })
        while #q > 0 do
            local b1, b2, ct = table.unpack(table.remove(q, 1))
            if st[b1 .. "," .. b2] == 1 then
                goto continue
            end
            if args.start_bucket == 1 then
                if b1 == 0 and b2 == args.bucket_two_capacity then
                    goto continue
                end
            end
            if args.start_bucket == 2 then
                if b2 == 0 and b1 == args.bucket_one_capacity then
                    goto continue
                end
            end
            if b1 == args.goal_volume then
                return {
                    moves = ct,
                    other_bucket_volume = b2,
                    goal_bucket_number = 1,
                }
            end

            if b2 == args.goal_volume then
                return {
                    moves = ct,
                    other_bucket_volume = b1,
                    goal_bucket_number = 2,
                }
            end
            st[b1 .. "," .. b2] = 1
            local diff = math.min(b1, args.bucket_two_capacity - b2)
            table.insert(q, { b1 - diff, b2 + diff, ct + 1 })
            table.insert(q, { 0, b2, ct + 1 })
            table.insert(q, { args.bucket_one_capacity, b2, ct + 1 })
            diff = math.min(args.bucket_one_capacity - b1, b2)
            table.insert(q, { b1 + diff, b2 - diff, ct + 1 })
            table.insert(q, { b1, 0, ct + 1 })
            table.insert(q, { b1, args.bucket_two_capacity, ct + 1 })
            ::continue::
        end
        error("impossible")
    end,
}
